systems/plants/RigidBodyManipulator.cpp

```diff
@@ -427,36 +438,68 @@ void RigidBodyManipulator::resize(int ndof, int num_featherstone_bodies, int num
 
 void RigidBodyManipulator::compile(void)
 {
+  /* todo:
+   - set joint limits (from drakejoint to rbm)
+   - add collision elements (and don't add them before, e.g. in constructModelmex)
+   */
+
+  // reorder body list to make sure that parents before children in the list
+  for (int i=0; i<bodies.size(); i++) {
+    while (bodies[i]->hasParent()) {
+      auto iter = find(bodies.begin()+i+1,bodies.end(),bodies[i]->parent);
+      if (iter==bodies.end()) break;
+      bodies[i].swap(*iter);
+    }
+  }
+
+  num_bodies = bodies.size();
+  num_dof = 0;
   num_velocities = 0;
   for (auto it = bodies.begin(); it != bodies.end(); ++it) {
     RigidBody& body = **it;
     if (body.hasParent()) {
+      body.dofnum = num_dof;
+      num_dof += body.getJoint().getNumPositions();
       body.velocity_num_start = num_velocities;
       num_velocities += body.getJoint().getNumVelocities();
     }
     else {
+      body.dofnum = 0;
       body.velocity_num_start = 0;
     }
   }
-  for (int i = 0; i < num_bodies; i++) {
+  for (int i=0; i<num_bodies; i++) {
+    bodies[i]->body_index = i;
     bodies[i]->setN(num_dof, num_velocities);
   }
 
+  B.resize(num_velocities,actuators.size());
+  B = MatrixXd::Zero(num_velocities,actuators.size());
+  for (int ia=0; ia<actuators.size(); ia++)
+    for (int i=0; i<actuators[ia].body->getJoint().getNumVelocities(); i++)
+      B(actuators[ia].body->velocity_num_start+i,ia) = actuators[ia].reduction;
+
+  resize(num_dof, NB, num_bodies, num_frames);
+
   for (int i=0; i<num_bodies; i++) {
     // precompute sparsity pattern for each rigid body
     bodies[i]->computeAncestorDOFs(this);
   }
+
   cached_q.resize(num_dof);
   cached_v.resize(num_velocities);
 
   initialized=true;
 }
 
-void RigidBodyManipulator::addCollisionElement(const int body_ind, const Matrix4d & T_elem_to_lnk, DrakeCollision::Shape shape, vector<double> params, string group_name)
+void RigidBodyManipulator::addCollisionElement(const int body_index, const Matrix4d & T_elem_to_lnk, DrakeCollision::Shape shape, vector<double> params, string group_name)
 {
-  bool is_static = (bodies[body_ind]->parent==0);
```

Comment:
this seemed like a bug to me.  why is it static if the parent is the world?  it should be static if the current body is the world.  that's how i've implemented it below - please tell me if I'm missing something.